### PR TITLE
[9.0.0][Formulas] Make all CRUD-related hook callbacks pass physical indexes to the formula engine.

### PR DESCRIPTION
**Pull request by jansiegel**
### Context
Make all CRUD-related hook callbacks pass physical indexes to the formula engine.

<sup>[skip changelog]</sup>

### How has this been tested?
Added a test case.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1. #8091 

### Affected project(s):
- [x] `handsontable`
- [ ] `@handsontable/angular`
- [ ] `@handsontable/react`
- [ ] `@handsontable/vue`

### CODE REVIEW

**Review by budnix — CHANGES_REQUESTED** [review]
It seems that not all changes are covered with the tests. When I reverted the login from the following hooks:
 * `onBeforeCreateRow`
 * `onBeforeCreateCol`
 * `onBeforeRemoveRow`
 * `onBeforeRemoveCol`
 * `onAfterCreateRow`
 * `onAfterCreateCol`
 * `onAfterRemoveCol`

The tests still pass. Can I ask you to write the missing tests?

**Review by aninde — APPROVED** [review]
🥇

**Comment by jansiegel** [comment]
> It seems that not all changes are covered with the tests. When I reverted the logic from the following hooks:
> 
> * `onBeforeCreateRow`
> * `onBeforeCreateCol`
> * `onBeforeRemoveRow`
> * `onBeforeRemoveCol`
> * `onAfterCreateRow`
> * `onAfterCreateCol`
> * `onAfterRemoveCol`
> 
> The tests still pass. Can I ask you to write the missing tests?

I split the test case into 8 separate cases - one covering each hook callback.

File: src/plugins/formulas/__tests__/formulas.spec.js [hunk 1]
```diff
@@ -2098,6 +2098,197 @@ describe('Formulas general', () => {
...
+            engine: HyperFormula
+          },
+          trimRows: [0, 1, 2, 3]
+        });
+
+        hot.setDataAtCell(2, 0, '{=TRANSPOSE(A1:B2)}');
+
+        hot.alter('remove_row', 3, 1);
+
+        expect(hot.getData().length).toEqual(4);
```
Comment (budnix) [inline src/plugins/formulas/__tests__/formulas.spec.js:2125]:
```suggestion
        expect(hot.countRows()).toBe(4);
```

File: src/plugins/formulas/__tests__/formulas.spec.js [hunk 2]
```diff
@@ -2098,6 +2098,197 @@ describe('Formulas general', () => {
...
+            engine: HyperFormula
+          },
+          manualColumnMove: [4, 5, 6, 7, 0, 1, 2, 3]
+        });
+
+        hot.setDataAtCell(0, 2, '{=TRANSPOSE(A1:B2)}');
+
+        hot.alter('remove_col', 3, 1);
+
+        expect(hot.getData().map(row => row.length)).toEqual([8, 8]);
```
Comment (budnix) [inline src/plugins/formulas/__tests__/formulas.spec.js:2144]:
```suggestion
        expect(hot.countCols()).toBe(8);
```

File: src/plugins/formulas/__tests__/formulas.spec.js [hunk 3]
```diff
@@ -2098,6 +2098,197 @@ describe('Formulas general', () => {
...
+            engine: HyperFormula
+          },
+          trimRows: [0, 1, 2, 3]
+        });
+
+        hot.setDataAtCell(2, 0, '{=TRANSPOSE(A1:B2)}');
+
+        hot.alter('insert_row', 3, 1);
+
+        expect(hot.getData().length).toEqual(4);
```
Comment (budnix) [inline src/plugins/formulas/__tests__/formulas.spec.js:2169]:
```suggestion
        expect(hot.countRows()).toBe(4);
```

File: src/plugins/formulas/__tests__/formulas.spec.js [hunk 4]
```diff
@@ -2098,6 +2098,197 @@ describe('Formulas general', () => {
...
+            engine: HyperFormula
+          },
+          manualColumnMove: [4, 5, 6, 7, 0, 1, 2, 3]
+        });
+
+        hot.setDataAtCell(0, 2, '{=TRANSPOSE(A1:B2)}');
+
+        hot.alter('insert_col', 3, 1);
+
+        expect(hot.getData().map(row => row.length)).toEqual([8, 8]);
```
Comment (budnix) [inline src/plugins/formulas/__tests__/formulas.spec.js:2188]:
```suggestion
        expect(hot.countCols()).toBe(8);
```

File: src/plugins/formulas/formulas.js [hunk 5]
```diff
@@ -491,12 +492,13 @@ export class Formulas extends BasePlugin {
...
    * @param {number} amount Amount of columns to be removed.
+   * @param {number[]} physicalColumns An array of physical columns removed from the data source.
    * @returns {*|boolean} If false is returned the action is canceled.
    */
-  onBeforeRemoveCol(col, amount) {
-    if (!this.engine.isItPossibleToRemoveColumns(this.sheetId, [col, amount])) {
-      return false;
-    }
+  onBeforeRemoveCol(col, amount, physicalColumns) {
+    return physicalColumns.every((physicalColumn) => {
```
Comment (budnix) [inline src/plugins/formulas/formulas.js:499]: Keep in mind that the return value overwrites the first argument. When the engine allows removing columns or rows the `undefined` should be returned.

Comment (jansiegel) [reply src/plugins/formulas/formulas.js:499]: Not `col`? Still, returning `true` should work just as well, right? https://github.com/handsontable/handsontable/blob/master/src/dataMap.js#L482

Comment (budnix) [reply src/plugins/formulas/formulas.js:499]: No, if you try listen to the changes e.q 
```js
beforeRemoveCol(col) {
  console.log('beforeRemoveCol', col);
},
```
with Formulas enabled you've got `beforeRemoveCol true` in the console.

Comment (jansiegel) [reply src/plugins/formulas/formulas.js:499]: True 👍

**Review by budnix — APPROVED** [review]
Thanks! Now everything is covered 👏

**Pull request merged**